### PR TITLE
fix(stash): skip empty pathspec stashes

### PR DESCRIPTION
**Pull request by jdx**
## Summary

- preflight path-limited stashes against the HEAD-to-worktree diff Git uses to reset paths
- skip `git stash push` when that patch is empty, before Git can reset the index and leave an orphaned stash
- cover both mode-only index changes and staged content whose worktree was restored to HEAD

Addresses the failure reported in https://github.com/jdx/hk/discussions/1281.

## Testing

- `mise run test:bats test/stash_empty_pathspec.bats`
- `mise run test:bats test/pre_commit_partial_staged.bats`
- `mise run test:cargo` (284 tests)
- `cargo fmt --check --manifest-path Cargo.toml`
- `mise run build`

`hk check --all` was also attempted. Its actionlint step could not run because this environment has no configured shellcheck shim; cargo-fmt completed successfully before that environment failure.

*AI-assisted — Tool: Codex; model: openai/gpt-5; version: unavailable.*

---

> [!NOTE]
> **Medium Risk**
> Changes pre-commit stash gating around index and staged state; incorrect diff handling could still skip needed stashes or leave edge cases, though behavior is narrowly scoped and covered by new BATS tests.
> 
> **Overview**
> Fixes pre-commit **git stash** runs that could **reset the index** and leave an orphaned stash when the pathspec had index changes but **no HEAD→worktree diff** (e.g. `git update-index --chmod`, or staged content with the worktree restored to HEAD).
> 
> Before `git stash push` with pathspecs, **`push_stash`** now runs `git diff --quiet HEAD -- <paths>`. Exit **0** means skip stashing (`Ok(None)`); **1** proceeds as before. Untracked-only cases still use the existing full-stash path when `HK_STASH_UNTRACKED` is set.
> 
> Adds **`test/stash_empty_pathspec.bats`** for mode-only staged commits, staged-vs-HEAD worktree mismatch, and untracked files alongside an empty stash patch.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 768ad30dd960af1bcdafca642a357d4e1084b5d0. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented empty Git stash operations from altering or losing staged changes.
  * Ensured staged file mode and content changes are preserved during commits.
  * Avoided leaving unnecessary stash entries after successful commits.

### CODE REVIEW

File: src/git.rs [hunk 1]
```diff
@@ -876,6 +876,37 @@ impl Git {
...
         }
+        // A path-limited stash decides what to reset from the HEAD-to-worktree diff,
+        // even though the paths above were selected from the index-to-worktree diff.
+        // If the former is empty (for example after `git update-index --chmod`), Git
+        // creates a stash entry and resets the index before failing to reverse-apply
+        // an empty patch. Avoid invoking Git in that destructive case.
+        if let Some(ref ts) = tracked_subset
+            && !ts.is_empty()
+            && (!*env::HK_STASH_UNTRACKED || status.untracked_files.is_empty())
+        {
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/git.rs:886]: Fixed in 768ad30d. The preflight now runs regardless of unrelated untracked files, and the new regression verifies the staged mode change commits, the untracked file remains intact, and no stash is left behind under both libgit2 settings.

*AI-assisted — Tool: Codex; model: openai/gpt-5; version: unavailable.*

File: src/git.rs [hunk 2]
```diff
@@ -876,6 +876,37 @@ impl Git {
...
+            match diff_status.code() {
+                Some(0) => return Ok(None),
+                Some(1) => {}
+                code => {
+                    return Err(eyre!(
+                        "git diff failed while checking stash pathspec with exit code {code:?}"
+                    ));
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/git.rs:908]: Fixed in 768ad30d. I reproduced that `--include-untracked` with tracked pathspecs does not protect this case, removed the untracked-file bypass, and added end-to-end coverage.

*AI-assisted — Tool: Codex; model: openai/gpt-5; version: unavailable.*

File: src/git.rs [hunk 3]
```diff
@@ -876,6 +876,36 @@ impl Git {
...
+        // an empty patch. Avoid invoking Git in that destructive case.
+        if let Some(ref ts) = tracked_subset
+            && !ts.is_empty()
+        {
+            let diff_status = Command::new("git")
+                .args([
+                    "diff",
+                    "--quiet",
+                    "--no-ext-diff",
+                    "--ignore-submodules",
```
[the remark this answers is not in the document]
Comment (jdx) [reply src/git.rs:892]: I do not think this should change. I reproduced the case with a dirty tracked submodule and an unrelated untracked file: `git diff --quiet HEAD -- sub` returns 1, but `git stash push --keep-index --include-untracked -- sub` prints `No local changes to save`, exits 0, leaves both the submodule dirt and untracked file untouched, and creates no stash. Git stash does not recurse into the submodule, so removing `--ignore-submodules` would not provide isolation; it would only make hk invoke a no-op stash and potentially mistake a pre-existing `stash@{0}` for one it created. The guard intentionally follows the stashable superproject diff here.

*AI-assisted — Tool: Codex; model: openai/gpt-5; version: unavailable.*
